nevergrad/functions/powersystems/core.py

```diff
@@ -0,0 +1,154 @@
+    """
+
+    def get_num_vars(self) -> List[Any]:
+        return [m.GetParamNumbers() for m in self.dam_managers]
+
+    def _simulate_power_system(self, input_x: np.ndarray) -> float:
+        dam_managers = self.dam_managers
+        x = list(input_x)
+        for a in dam_managers:
+            assert(len(x) >= a.GetParamNumbers())
+            a.SetParams(np.array(x[:a.GetParamNumbers()]))
+            x = x[a.GetParamNumbers():]
+        assert(len(x) == 0)
+        N = int(self.N)
+        # Assume empty initial stocks.
+        stocks = [0.] * N   
+        # Nonsense delays.
+        delay = [np.cos(i) for i in range(N)]
+        cost = 0.
+        # Loop on time steps.
+        consumption = 0.
+        hydro_prod_per_time_step: List[float] = []
+        consumption_per_time_step: List[float] = []
+        for t in range(int(365*24*self.number_of_years)):
+    
+            # Rain
+            for i in range(N):
+                stocks[i] += 0.5*(1.+np.cos(2*pi*t/(24*365) + delay[i])) * np.random.rand()
+            # Consumption model.
+            base_consumption = (self.constant_to_year_ratio*self.year_to_day_ratio 
+                    +0.5*self.year_to_day_ratio*(1.+np.cos(2*pi*t/(24*365))) + 0.5*(1.+np.cos(2*pi*t/24)))
+            if t == 0:
+                consumption = base_consumption
+            else:
+                consumption = max(0., consumption + self.consumption_noise*(np.random.rand()-.5) + self.back_to_normal * (base_consumption - consumption))
+            consumption_per_time_step += [consumption]
+            # Water usage.
+            needed = consumption
+    
+            # Setting inputs for all agents.
+            base_x = [np.cos(2*pi*t/24.), np.sin(2*pi*t/24.), np.cos(2*pi*t/(365*24)), np.sin(2*pi*t/(365*24)), needed, self.average_consumption, self.year_to_day_ratio, self.constant_to_year_ratio, self.back_to_normal, self.consumption_noise]
+            x = list(base_x + self.thermal_power_capacity + self.thermal_power_prices + stocks)
+    
+            # Prices as a decomposition tool!
+            price: List[float] = [a.GetOutput(np.array(x))[0][0] for a in dam_managers]
+            volume: List[float] = [s for s in stocks]
+            dam_index: List[int] = list(range(len(price)))
+            price += self.thermal_power_prices
+            volume += self.thermal_power_capacity
+            dam_index += [-1] * len(price)
+            
+            assert(len(price) == N + self.num_thermal_plants)
+            hydro_prod: List[float] = [0.] * N
+            for i in range(len(price)):
+                if needed == 0:
+                    break
+                production = min(volume[i], needed)
+                if dam_index[i] >= 0:
+                    hydro_prod[dam_index[i]] += production
+                    stocks[dam_index[i]] -= production
+                    assert(stocks[dam_index[i]] >= -1e-7)
+                else:
+                    cost += production * price[i]
```

Comment:
That never happens.

Reply:
Should happen, if "needed" is too high ?

Reply:
What I mean is that dam_index is defined as `list(range(len(price))) + [-1] * len(price)` in l108 and l111.
You loop on i in range(len(price)) so dam_index[i] will take the values of range(len(price)) (by the way it would be clearer if you were to just do for dam_index in range(len(price)) instead of this loop).
Then your test `if dam_index[i] >=0` will always return true and you will never execute what is in else.

Maybe that's not the test you actually want to do ?

Reply:
oh ok sorry I didn't read your update before reading this. You corrected it !